tests/unit/rest/test_client.py

```diff
@@ -44,3 +50,102 @@ def test_obsolete_exception_twiliotaskrouterclient(self):
+        self.http_client_constructor_mock.return_value = self.http_client_mock
+
+        self.client = Client('username', 'password')
+
+    def tearDown(self):
+        self.http_client_patcher.stop()
+
+    def validate_url(self, method, url, **kwargs):
+        if url == self.request_mock.url:
+            return Response(200, 'test response')
+
+        return Response(
+            500,
+            'Incorrect url: expected {expected}, got {returned}'.format(
+                expected=self.request_mock.url,
+                returned=url)
+        )
+
+    def validate_success_response(self, response):
+        self.assertEqual(200, response.status_code)
+        self.assertEqual('test response', response.content)
+
+    def test_set_client_edge_default_region(self):
+        self.request_mock.url = 'https://api.edge.us1.twilio.com'
+
+        self.client.edge = 'edge'
+        response = self.client.request('GET', 'https://api.twilio.com')
+
+        self.validate_success_response(response)
+
+    def test_set_client_region(self):
+        self.request_mock.url = 'https://api.region.twilio.com'
+
+        self.client.region = 'region'
+        response = self.client.request('GET', 'https://api.twilio.com')
+
+        self.validate_success_response(response)
+
+    def test_set_uri_region(self):
+        self.request_mock.url = 'https://api.region.twilio.com'
+
+        response = self.client.request('GET', 'https://api.region.twilio.com')
+
+        self.validate_success_response(response)
+
+    def test_set_client_edge_region(self):
+        self.request_mock.url = 'https://api.edge.region.twilio.com'
+
+        self.client.edge = 'edge'
+        self.client.region = 'region'
+        response = self.client.request('GET', 'https://api.twilio.com')
+
+        self.validate_success_response(response)
+
+    def test_set_client_edge_uri_region(self):
+        self.request_mock.url = 'https://api.edge.region.twilio.com'
+
+        self.client.edge = 'edge'
+        response = self.client.request('GET', 'https://api.region.twilio.com')
+
+        self.validate_success_response(response)
+
+    def test_set_client_region_uri_edge_region(self):
+        self.request_mock.url = 'https://api.edge.region.twilio.com'
+
+        self.client.region = 'region'
+        response = self.client.request('GET', 'https://api.edge.uriRegion.twilio.com')
+
+        self.validate_success_response(response)
+
+    def test_set_client_edge_uri_edge_region(self):
+        self.request_mock.url = 'https://api.edge.region.twilio.com'
+
+        self.client.edge = 'edge'
+        response = self.client.request('GET', 'https://api.uriEdge.region.twilio.com')
+
+        self.validate_success_response(response)
+
+    def test_set_uri_edge_region(self):
+        self.request_mock.url = 'https://api.edge.region.twilio.com'
+
+        response = self.client.request('GET', 'https://api.edge.region.twilio.com')
```

Comment:
Should add a test with a `.` in the query params/or fragment that would break if not for `urlparse` logic.

Reply:
Confirmed that the test I added breaks without `urlparse` and succeeds with it.